fix: typescript definitions

sublimator opened the pull request:
This should make the typescript definitions at least useable. Currently they seem to be quite broken (per #69, #67 )  

Closes #54 
Closes #47 
Closes #69 
Closes #67 
Closes #62 

- Adds examples/typescript and build in CI step
- Fixes definitions (trying to set defaults, chatStream returning wrong type)
- Add signal (per #68) to typescript definitions 
- Fix interfaces (non-exhaustively) per: https://docs.mistral.ai/redocusaurus/plugin-redoc-0.yaml

sublimator commented:
@fuegoio

Please look :)

fuegoio commented on src/client.d.ts:
```diff
@@ -1,194 +1,177 @@
+export interface Message {
+  role: string;
+  // TODO: this is not specified at https://docs.mistral.ai/redocusaurus/plugin-redoc-0.yaml
+  name?: string,
```
Normally there is no `name` on the message, I think we can remove it.

fuegoio commented on src/client.d.ts:
```diff
@@ -1,194 +1,177 @@
+  function: FunctionCall;
+}
+
+export enum ResponseFormats {
```
I think we should also remove the enums from the types and change them to types. A simple example like this does not work because the enums are TS only:

```ts
import MistralClient, { ResponseFormats } from "@mistralai/mistralai";

const chatResponse = await mistralClient.chat({
    model: "mistral-large",
    temperature: 0.1,
    responseFormat: { type: ResponseFormats.json_object },
})`
```

This raises the error `'mistralai/src/client.js' does not provide an export named 'ResponseFormats'`

sublimator commented:
>
> I suppose someone put "name" in there for OpenAI compat reasons at some
point, though it is just misleading, and also not necessary

>
>> Message ID: ***@***.***>
>>
>

sublimator commented:
Oh right, yeah, the enums aren't even defined at runtime are they. I added
literals (for ergonomics) and left the enums in there and moved on. They
are completely broken though aren't they.

A good argument to move to TS :)

On Tue, 7 May 2024, 2:51 pm Nicholas Dudfield, ***@***.***>
wrote:

> >
> > I suppose someone put "name" in there for OpenAI compat reasons at some
> point, though it is just misleading, and also not necessary
>
> >
> >> Message ID: ***@***.***>
> >>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mistralai/client-js/pull/72#issuecomment-2097669949>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEAHG6QYRPI5CS72RB4DX3ZBCBWLAVCNFSM6AAAAABHKCIQK6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJXGY3DSOJUHE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

fuegoio commented:
> Oh right, yeah, the enums aren't even defined at runtime are they. I added literals (for ergonomics) and left the enums in there and moved on. They are completely broken though aren't they. A good argument to move to TS :)

In a `d.ts` file they do not make much sense no. I approve moving to TS to have this kind of nice feature.

sublimator commented:
@fuegoio 

Anything else? I'll look at a full TS port when I have more time

sublimator commented:
Oh, wait, that's not working, enums still referenced, but no errors, hrmmm

sublimator commented on src/client.d.ts:
```diff
@@ -1,194 +1,160 @@
-    export interface ToolCalls {
-        id: 'null';
-        type: ToolType = ToolType.function;
-        function: FunctionCall;
```
What happened here? :)

fuegoio replied:
What do you mean?

sublimator replied:
```
id: 'null';
type: ToolType = ToolType.function;
```

As you mentioned, the enums don't work at runtime, and `null` is set to match the string null, not the null *type* (which is actually not appropriate either). Also, it's trying to set a default (`ToolType = ToolType.function;`) inside a `.d.ts` file, which I don't believe actually works. 

More a rhetorical question than serious :) Wondering how the ..

fuegoio replied:
Yeah the `null` string is expected, as before our API change the id returned by the API was always `'null'` (the string). Now it is function call ids yes. The default stuff well I'm not sure ahah

sublimator commented:
>  I'll look at a full TS port when I have more time

In a separate PR, to be clear

fuegoio reviewed (approved):
Thanks a lot @sublimator!
